File: backend/src/domain/repository/IMentorRepository.ts
import { MentorProfile } from '../entities/MentorProfile';
import { MentorExpertise } from '../entities/MentorExpertise';
import { MentorAvailability } from '../entities/MentorAvailability';
import { MentorshipSession } from '../entities/MentorshipSession';
import { SessionFeedback } from '../entities/SessionFeedback';

/**
 * Repository interface for mentorship operations
 */
export interface IMentorRepository {
  // Mentor Profile Management
  createMentorProfile(
    profileData: Partial<MentorProfile>,
  ): Promise<MentorProfile>;
  updateMentorProfile(
    memberId: number,
    profileData: Partial<MentorProfile>,
  ): Promise<MentorProfile>;
  findMentorProfileByMemberId(memberId: number): Promise<MentorProfile | null>;
  findAllMentorProfiles(
    organizationId: number,
    page: number,
    limit: number,
    filters?: { isApproved?: boolean },
  ): Promise<{ mentors: MentorProfile[]; total: number }>;
  approveMentorProfile(memberId: number): Promise<MentorProfile>;
  updateMentorRating(
    memberId: number,
    newAvgRating: number,
    totalSessions: number,
  ): Promise<MentorProfile>;

  // Mentor Expertise Management
  createExpertise(
    expertiseData: Partial<MentorExpertise>,
  ): Promise<MentorExpertise>;
  updateExpertise(
    expertiseId: number,
    expertiseData: Partial<MentorExpertise>,
  ): Promise<MentorExpertise>;
  deleteExpertise(expertiseId: number): Promise<boolean>;
  findExpertiseByMentor(mentorMemberId: number): Promise<MentorExpertise[]>;
  searchMentorsByExpertise(
    topic: string,
    page: number,
    limit: number,
  ): Promise<{ mentors: MentorProfile[]; total: number }>;

  // Availability Management
  createAvailability(
    availabilityData: Partial<MentorAvailability>,
  ): Promise<MentorAvailability>;
  updateAvailability(
    availabilityId: number,
    availabilityData: Partial<MentorAvailability>,
  ): Promise<MentorAvailability>;
  deleteAvailability(availabilityId: number): Promise<boolean>;
  findAvailabilityById(
    availabilityId: number,
  ): Promise<MentorAvailability | null>;
  findAvailabilitiesByMentor(
    mentorMemberId: number,
    startDate?: Date,
    endDate?: Date,
  ): Promise<MentorAvailability[]>;
  findAvailableSlots(
    mentorMemberId: number,
    startDate: Date,
    endDate: Date,
  ): Promise<MentorAvailability[]>;
  updateAvailabilityStatus(
    availabilityId: number,
    status: string,
  ): Promise<MentorAvailability>;

  // Session Management
  createSession(
    sessionData: Partial<MentorshipSession>,
  ): Promise<MentorshipSession>;
  updateSession(
    sessionId: number,
    sessionData: Partial<MentorshipSession>,
  ): Promise<MentorshipSession>;
  cancelSession(sessionId: number): Promise<MentorshipSession>;
  completeSession(sessionId: number): Promise<MentorshipSession>;
  findSessionById(sessionId: number): Promise<MentorshipSession | null>;
  findSessionsByMentor(
    mentorMemberId: number,
    page: number,
    limit: number,
  ): Promise<{ sessions: MentorshipSession[]; total: number }>;
  findSessionsByMentee(
    menteeMemberId: number,
    page: number,
    limit: number,
  ): Promise<{ sessions: MentorshipSession[]; total: number }>;
  findUpcomingSessions(memberId: number): Promise<MentorshipSession[]>;

  // Feedback Management
  createFeedback(
    feedbackData: Partial<SessionFeedback>,
  ): Promise<SessionFeedback>;
  updateFeedback(
    feedbackId: number,
    feedbackData: Partial<SessionFeedback>,
  ): Promise<SessionFeedback>;
  findFeedbackBySession(sessionId: number): Promise<SessionFeedback | null>;
  findFeedbacksByMentor(
    mentorMemberId: number,
    page: number,
    limit: number,
    publicOnly?: boolean,
  ): Promise<{ feedbacks: SessionFeedback[]; total: number }>;

  // Statistics
  getMentorStatistics(mentorMemberId: number): Promise<{
    totalSessions: number;
    completedSessions: number;
    averageRating: number;
    totalFeedbacks: number;
  }>;
}

File: backend_java/src/main/java/com/service/backend/domain/repository/IMentorRepository.java
package com.service.backend.domain.repository;

import com.service.backend.domain.entity.MentorProfile;
import com.service.backend.domain.entity.MentorExpertise;
import com.service.backend.domain.entity.MentorAvailability;
import com.service.backend.domain.entity.MentorshipSession;
import com.service.backend.domain.entity.SessionFeedback;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Map;

/**
 * Repository interface for mentorship operations
 */
public interface IMentorRepository {

    // Mentor Profile Management
    Mono<MentorProfile> createMentorProfile(MentorProfile profileData);
    Mono<MentorProfile> updateMentorProfile(Long memberId, MentorProfile profileData);
    Mono<MentorProfile> findMentorProfileByMemberId(Long memberId);
    Mono<Map<String, Object>> findAllMentorProfiles(Long organizationId, int page, int limit, Map<String, Object> filters);
    Mono<MentorProfile> approveMentorProfile(Long memberId);
    Mono<MentorProfile> updateMentorRating(Long memberId, BigDecimal newAvgRating, Integer totalSessions);

    // Mentor Expertise Management
    Mono<MentorExpertise> createExpertise(MentorExpertise expertiseData);
    Mono<MentorExpertise> updateExpertise(Long expertiseId, MentorExpertise expertiseData);
    Mono<Boolean> deleteExpertise(Long expertiseId);
    Flux<MentorExpertise> findExpertiseByMentor(Long mentorMemberId);
    Mono<Map<String, Object>> searchMentorsByExpertise(String topic, int page, int limit);

    // Availability Management
    Mono<MentorAvailability> createAvailability(MentorAvailability availabilityData);
    Mono<MentorAvailability> updateAvailability(Long availabilityId, MentorAvailability availabilityData);
    Mono<Boolean> deleteAvailability(Long availabilityId);
    Mono<MentorAvailability> findAvailabilityById(Long availabilityId);
    Flux<MentorAvailability> findAvailabilitiesByMentor(Long mentorMemberId, LocalDateTime startDate, LocalDateTime endDate);
    Flux<MentorAvailability> findAvailableSlots(Long mentorMemberId, LocalDateTime startDate, LocalDateTime endDate);
    Mono<MentorAvailability> updateAvailabilityStatus(Long availabilityId, String status);

    // Session Management
    Mono<MentorshipSession> createSession(MentorshipSession sessionData);
    Mono<MentorshipSession> updateSession(Long sessionId, MentorshipSession sessionData);
    Mono<MentorshipSession> cancelSession(Long sessionId);
    Mono<MentorshipSession> completeSession(Long sessionId);
    Mono<MentorshipSession> findSessionById(Long sessionId);
    Mono<Map<String, Object>> findSessionsByMentor(Long mentorMemberId, int page, int limit);
    Mono<Map<String, Object>> findSessionsByMentee(Long menteeMemberId, int page, int limit);
    Flux<MentorshipSession> findUpcomingSessions(Long memberId);

    // Feedback Management
    Mono<SessionFeedback> createFeedback(SessionFeedback feedbackData);
    Mono<SessionFeedback> updateFeedback(Long feedbackId, SessionFeedback feedbackData);
    Mono<SessionFeedback> findFeedbackBySession(Long sessionId);
    Mono<Map<String, Object>> findFeedbacksByMentor(Long mentorMemberId, int page, int limit, Boolean publicOnly);

    // Statistics
    Mono<Map<String, Object>> getMentorStatistics(Long mentorMemberId);
}
